src/Protocol/ForgeHandshake.cpp

```diff
@@ -0,0 +1,366 @@
+
+	size_t ModCount = ServerMods.size();
+
+	Buf.WriteBEInt8(Discriminator::ModList);
+	Buf.WriteVarInt32(static_cast<UInt32>(ModCount));
+	for (auto & item: ServerMods)
+	{
+		const AString & name = item.first;
+		const AString & version = item.second;
+		Buf.WriteVarUTF8String(name);
+		Buf.WriteVarUTF8String(version);
+	}
+	AString ServerModList;
+	Buf.ReadAll(ServerModList);
+
+	m_Client->SendPluginMessage("FML|HS", ServerModList);
+}
+
+
+
+
+void cForgeHandshake::HandleHandshakeAck(cClientHandle * a_Client, const char * a_Data, size_t a_Size)
+{
+	if (a_Size != 2)
+	{
+		LOG("Unexpected HandshakeAck packet length: " SIZE_T_FMT "", a_Size);
+		SetError();
+		return;
+	}
+
+	int Phase = a_Data[1];
+	LOGD("Received client HandshakeAck with phase=%d", Phase);
+
+	switch (Phase)
+	{
+		case ClientPhase::WAITINGSERVERDATA:
+		{
+			cByteBuffer Buf(1024);
+			Buf.WriteBEInt8(Discriminator::RegistryData);
+
+			// TODO: send real registry data
+			bool HasMore = false;
+			AString RegistryName = "potions";
+			UInt32 NumIDs = 0;
+			UInt32 NumSubstitutions = 0;
+			UInt32 NumDummies = 0;
+
+			Buf.WriteBool(HasMore);
+			Buf.WriteVarUTF8String(RegistryName);
+			Buf.WriteVarInt32(NumIDs);
+			Buf.WriteVarInt32(NumSubstitutions);
+			Buf.WriteVarInt32(NumDummies);
+
+			AString RegistryData;
+			Buf.ReadAll(RegistryData);
+			m_Client->SendPluginMessage("FML|HS", RegistryData);
+			break;
+		}
+
+		case ClientPhase::WAITINGSERVERCOMPLETE:
+		{
+			LOGD("Client finished receiving registry data; acknowledging");
+
+			AString Ack;
+			Ack.push_back(Discriminator::HandshakeAck);
+			Ack.push_back(ServerPhase::WAITINGCACK);
+			m_Client->SendPluginMessage("FML|HS", Ack);
+			break;
+		}
+
+		case ClientPhase::PENDINGCOMPLETE:
+		{
+			LOGD("Client is pending completion; sending complete ack");
+
+			AString Ack;
+			Ack.push_back(Discriminator::HandshakeAck);
+			Ack.push_back(ServerPhase::COMPLETE);
+			m_Client->SendPluginMessage("FML|HS", Ack);
+
+			// Now finish logging in
+			m_Client->FinishAuthenticate(m_Name, m_UUID, m_Properties);
+			break;
+		}
+	}
+}
+
+
+
+
+
+void cForgeHandshake::DataReceived(cClientHandle * a_Client, const char * a_Data, size_t a_Size)
+{
+	if (!m_IsForgeClient)
+	{
+		LOG("Received unexpected Forge data from non-Forge client (" SIZE_T_FMT " bytes)", a_Size);
+		return;
+	}
+	LOGD("Received Forge data: " SIZE_T_FMT " bytes: %s", a_Size, a_Data);
+
+	if (a_Size <= 1)
+	{
+		LOG("Received unexpectedly short Forge data (" SIZE_T_FMT " bytes)", a_Size);
+		return;
+	}
+
+	int Discriminator = a_Data[0];
+
+	switch (Discriminator)
+	{
+		case Discriminator::ClientHello:
+			HandleClientHello(a_Client, a_Data, a_Size);
+			break;
+
+		case Discriminator::ModList:
+			HandleModList(a_Client, a_Data, a_Size);
+			break;
+
+		case Discriminator::HandshakeAck:
+			HandleHandshakeAck(a_Client, a_Data, a_Size);
+			break;
+
+		default:
+		{
+			LOG("Unexpected Forge packet %d received", Discriminator);
+			SetError();
```

Comment:
Oh, I thought you'd just make it a `void SetError(const AString &)` and use `Printf` for the formatting, if needed.

This way is not really good, because the logging function is called twice, so it emits two messages, with two timestamps.